presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestCanonicalizeExpressionRewriter.java

```diff
@@ -39,8 +44,23 @@ public void testRewriteYearExtract()
         assertRewritten("EXTRACT(YEAR FROM '2017-07-20')", "year('2017-07-20')");
     }
 
-    private static void assertRewritten(String from, String to)
+    @Test
+    public void testRewriteMapConcatAndArrayConcat()
     {
-        assertEquals(canonicalizeExpression(PlanBuilder.expression(from)), PlanBuilder.expression(to));
+        assertRewritten(
+                "map_concat(map(ARRAY[1,2], ARRAY['a','b']), map_concat(map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f'])))",
+                "map_concat(map(ARRAY[1,2], ARRAY['a','b']), map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f']))");
+
+        assertRewritten(
+                "map(array_concat(array_concat(ARRAY[1]),array_concat(ARRAY[2])), ARRAY['a','b'])",
+                "map(array_concat(ARRAY[1],ARRAY[2]), ARRAY['a','b'])");
+
+        assertRewritten(
+                "array_concat(array[1,2],array[3,4], array_concat(array[5,6],array_concat(array[7,8])))",
+                "array_concat(array[1,2],array[3,4],array[5,6],array[7,8])");
+
+        assertRewritten(
+                "map_concat(map(array_concat(array_concat(ARRAY[1]),array_concat(ARRAY[2])), ARRAY['a','b']), map_concat(map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f'])))",
+                "map_concat(map(array_concat(ARRAY[1],ARRAY[2]), ARRAY['a','b']), map(ARRAY[3,4], ARRAY['c','d']), map(ARRAY[5,6], ARRAY['e', 'f']))");
```

Comment:
We should also have some negative tests that don't change the expression.

Reply:
made changes